velox/functions/sparksql/DateTimeFunctions.h

```diff
@@ -39,6 +39,34 @@ Expected<std::shared_ptr<DateTimeFormatter>> getDateTimeFormatter(
           std::string_view(format.data(), format.size()));
   }
 }
+
+// Initialize `formatter_` based on the provided format string. It determines
+// formatter type based on the `legacyFormatter` flag. The legacy formatter
+// returns false for an invalid format, while the Joda formatter throws a
+// Velox user error.
+//
+// @param format The format string to be used for initializing the formatter.
+// @param legacyFormatter Determines formatter type.
+// @return An optional shared pointer to a DateTimeFormatter. If the formatter
+// initialization fails and the legacy formatter is used, std::nullopt is
+// returned.
+std::optional<std::shared_ptr<DateTimeFormatter>> initializeFormatter(
```

Comment:
Could we use 'std::shared_ptr<DateTimeFormatter>' as return type and return nullptr for error case?

Reply:
@rui-mo I think `std::optional` allows the caller to explicitly check for the presence of a valid formatter. Could you please provide more details on why you prefer using 'std::shared_ptr'? Thanks!

Reply:
> std::optional allows the caller to explicitly check for the presence of a valid formatter.

I assume the information about its validity is contained in the pointer itself. Just wondering if we need another layer, thanks.